### PR TITLE
[Android] Toggle window

### PR DESCRIPTION
**Pull request by tapanmodh**
Resolves https://github.com/brave/brave-browser/issues/35106

## Submitter Checklist:

- [x] I confirm that no [security/privacy review is needed](https://github.com/brave/brave-browser/wiki/Security-reviews) and no other type of reviews are needed, or that I have [requested](https://github.com/brave/reviews/issues/new/choose) them
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [ ] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [x] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [x] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

## Test Plan:

| Screenshot_20240422_192234 | Screenshot_20240422_200941 | Screenshot_20240419_155636 |
| ------------- | ------------- | ------------- |

### CODE REVIEW

**Comment by tapanmodh** [comment]
| Screenshot_20240401_140719 | Screenshot_20240401_140829 |
| ------------- | ------------- |
| Screenshot 1 is when multiple windows are already there and user tries to disable windows. | Screenshot 2 is when user tap on merge windows from screenshot 1. It opens chromium's move to other window dialog. |

cc: @deeppandya @timchilds @anthonypkeane

**Comment by deeppandya** [comment]
> @tapanmodh @deeppandya re: screenshot 1.
> 
> * Header: I would change "MultiWindows" to "Multiple Windows"
> * Description: I would re-word this as follows:
> 
> "Please merge tabs to another window or close other windows before disabling the windows feature."
> 
> @rmcfadden3 WDYT?

@timchilds Looks good to me.

**Comment by rmcfadden3** [comment]
Agreed on header of `Multiple windows detected`
For description, might suggest a slight tweak to 'Please merge tabs to another Brave window, or close other Brave windows, before disabling the multiple-windows feature.`

File: browser/ui/android/strings/android_brave_strings.grd [hunk 1]
```diff
@@ -3932,6 +3932,27 @@ If you don't accept this request, VPN will not reconnect and your internet conne
     <message name="IDS_REWARDS_TOS_UPDATE_LINK_TEXT" desc="Terms of service text for rewards panel">
       See <ph name="TOS_START">&lt;tos&gt;</ph>Brave Rewards Terms of Service<ph name="TOS_END">&lt;/tos&gt;</ph>
     </message>
+    <message name="IDS_ENABLE_WINDOWS_TITLE" desc="Title for preference that enables or disables windows.">
+      Enable Windows
```
Comment (mkarolin) [inline browser/ui/android/strings/android_brave_strings.grd:3936]: Suggestion: Enable multiple windows

File: browser/ui/android/strings/android_brave_strings.grd [hunk 2]
```diff
@@ -3932,6 +3932,27 @@ If you don't accept this request, VPN will not reconnect and your internet conne
...
+    </message>
+    <message name="IDS_ENABLE_WINDOWS_SUMMARY" desc="Summary for preference that enables or disables windows.">
+      Open up to 5 windows at a time and move tabs from one window to another
+    </message>
+    <message name="IDS_MULTI_WINDOWS_DIALOG_TITLE" desc="Title for multiwindows detected dialog.">
+      Multiple Windows detected
```
Comment (mkarolin) [inline browser/ui/android/strings/android_brave_strings.grd:3942]: Needs consistent capitalization: either "Multiple windows detected" or "Multiple Windows Detected"

File: browser/ui/android/strings/android_brave_strings.grd [hunk 3]
```diff
@@ -3932,6 +3932,27 @@ If you don't accept this request, VPN will not reconnect and your internet conne
...
+    </message>
+    <message name="IDS_MULTI_WINDOWS_DIALOG_SUMMARY" desc="Summary for multiwindows detected dialog.">
+      Please merge tabs to another Brave window, or close other Brave windows, before disabling the multiple-windows feature.
+    </message>
+    <message name="IDS_MERGE_WINDOWS" desc="Button text of merge windows for multiwindows dialog">
+      Merge Windows
+    </message>
+    <message name="IDS_CLOSE_WINDOWS" desc="Button text of close windows for multiwindows dialog">
+      Close Windows
```
Comment (mkarolin) [inline browser/ui/android/strings/android_brave_strings.grd:3951]: Not sure what our capitalization is on buttons. Does the word "Windows" need to be capitalized here?

File: browser/ui/android/strings/android_brave_strings.grd (hunk 2 above)
Comment (tapanmodh) [reply browser/ui/android/strings/android_brave_strings.grd:3942]: My bad. I'll update it

File: android/java/org/chromium/chrome/browser/multiwindow/BraveMultiInstanceManagerApi31.java [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+                                            Snackbar.UMA_UNKNOWN)
+                                    .setAction(
+                                            braveActivity
+                                                    .getResources()
+                                                    .getString(R.string.merge_windows),
+                                            null)
+                                    .setSingleLine(false)
+                                    .setDuration(10000);
+                    SnackbarManager snackbarManager =
+                            SnackbarManagerProvider.from(braveActivity.getWindowAndroid());
```
Comment (SergeyZhukovsky) [inline android/java/org/chromium/chrome/browser/multiwindow/BraveMultiInstanceManagerApi31.java:96]: you can get `WindowAndroid` from
```
mTabModelOrchestratorSupplier
                .get()
                .getTabModelSelector()
                .getCurrentModel().get()
                .getWindowAndroid()
```

File: android/java/org/chromium/chrome/browser/multiwindow/BraveMultiInstanceManagerApi31.java [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+            TabModelSelector selector =
+                    TabWindowManagerSingleton.getInstance().getTabModelSelectorById(mInstanceId);
+            if (selector.getTotalTabCount() == 0) {
+                closeInstance(mInstanceId, INVALID_TASK_ID);
+            }
+            if (MultiWindowUtils.getInstanceCount() == 1) {
+                BraveMultiWindowUtils.updateEnableWindows(false);
+            } else {
+                try {
+                    BraveActivity braveActivity = BraveActivity.getBraveActivity();
```
Comment (SergeyZhukovsky) [inline android/java/org/chromium/chrome/browser/multiwindow/BraveMultiInstanceManagerApi31.java:70]: could we use Activity that we pass in the constructor? We should use `BraveActivity.getBraveActivity()` as a last resort if there is no Activity associated with a current control. You have an access to `mActivity` super's field.

Comment (tapanmodh) [reply android/java/org/chromium/chrome/browser/multiwindow/BraveMultiInstanceManagerApi31.java:70]: sure, I'll update it

File: browser/ui/android/strings/android_brave_strings.grd (hunk 1 above)
Comment (tapanmodh) [reply browser/ui/android/strings/android_brave_strings.grd:3936]: @timchilds @rmcfadden3 wdyt?

**Review by SergeyZhukovsky — APPROVED** [review]
++

**Review by deeppandya — APPROVED** [review]
LGTM

**Review by mkarolin — APPROVED** [review]
`strings`++

**Comment by bsclifton** [comment]
@tapanmodh can you update https://github.com/brave/brave-browser/issues/35106 and have a test plan that covers the cases?
1. default setting (off, right?)
2. default setting if person has multiple windows already
3. multiple windows needing to merge
4. feature off then turning on - what happens

May be other ones you think are important. But you can edit top post of https://github.com/brave/brave-browser/issues/35106 if you're able and then we can link to there under `Test plan` here. Thanks! 😄

**Comment by Uni-verse** [comment]
In progress on `Samsung Galaxy Tab S7` using version: 

```
Brave	1.67.73 Chromium: 125.0.6422.26 (Official Build) canary (64-bit) 
Revision	76e7d97680074902ad7bde84f0a1a6c392a4ca6a
OS	Android 13; Build/TP1A.220624.014; 33; REL
```

Use Case 1: New Install

- Ensured `New Window` is not shown in the menu.
- Ensured `Enable multiple windows` setting appears in `Appearance` menu and is defaulted to off for new users.
- Ensured enabling the setting will show the `New Window` option in the menus. 

_Disabled (Default)_

Setting | Homepage menu | Tab menu (URL) | Private tab menu 
-|-|-|-
Screenshot 2024-05-08 at 4 24 47 PM | Screenshot 2024-05-08 at 4 36 52 PM | Screenshot 2024-05-08 at 4 37 38 PM | Screenshot 2024-05-08 at 4 39 33 PM

_Enabled_

Example | Example | Example
-|-|-
Screenshot 2024-05-08 at 4 41 01 PM | Screenshot 2024-05-08 at 4 41 08 PM | Screenshot 2024-05-08 at 4 41 14 PM

Use Case 2: Upgrade from 1.66 -> 1.67 (Single window)

- Ensured that `New Window` is not shown in menus  when upgrading with a single window open.
- Ensured that enabling `Enable multiple windows` will show the feature in the menus.
- Ensured opening another window will enable multi-window mode.

Example | Example | Example
-|-|-
Screenshot 2024-05-08 at 6 27 53 PM | Screenshot 2024-05-08 at 6 27 59 PM | Screenshot 2024-05-08 at 6 28 43 PM

Use Case 3: Upgrade from 1.66 -> 1.67 (Multi-window)

- Ensured enabling `Enable multiple windows` is on by default for users who upgrade to 1.67.x
- Ensured `New window`, `Manage windows` and `Move to other window` are shown in the menu

Example | Example 
-|-
Screenshot 2024-05-08 at 5 49 08 PM | Screenshot 2024-05-08 at 4 42 12 PM

**Pull request merged**